src/2025h2/comprehensive-niche-checks.md

```diff
@@ -0,0 +1,118 @@
+`std::ptr::NonNull`. These so called *niches* are used for optimizations, for
+example when being used together with `Option`, where `Option<NonNull<T>>` has
+the same size as `*mut T`.
+
+In safe Rust it is never possible to construct such values, but that is
+different for unsafe Rust or FFI. E.g. with Rusts growing interop with C (e.g.
+in the Linux Kernel) and C++ it makes sense to expose such types over the
+language boundary. That can make for situations where a valid C or C++ value
+becomes an invalid Rust value.
+
+### The status quo
+
+While [Miri](https://github.com/rust-lang/miri) exists and does a great job at
+catching various types of UB in unsafe Rust code, it has the downside of only
+working on pure Rust code. Extern functions can not be called and a mixed
+language binary is unable to be executed in Miri.
+
+We already added support for checking enum discriminants which is somewhat related,
+but this does not cover general niches.
+
+At the same time we struggled where to inject the enum discriminant check. If
+we e.g. inject the check for every use (copy, move) of an enum, we already
+observe UB for invalid enums. E.g. for enums with just two variants and a
+representation of u8, the rust compiler uses an `i1` in the LLVM IR to represent the
+discriminant value ([godbolt](https://godbolt.org/z/Mf977hMsr) example). But this
+means we essentially have no way of inserting a check for the value at the
+moment. Any time we try to convert from an invalid value (e.g. 42), the compiler
+will insert a `trunc` to generate the correct discriminant backing value.
+With these optimizations enabled, we can never check for valid enum
+discriminants.
+
+Similar, any check for an argument marked `nonnull` will be optimized out by the
+compiler (which is the whole idea of the attribute).
+
+A solution to this could be some sort of "checked-load" inserts validity checks
+when values are loaded from memory. Alternatively we could insert checks for
+function arguments at the start of a function (e.g. by inserting a 
+"shadow block") that check the values for validity before they are used.
+
+### The next 6 months
+
+Within the next six months we would like to do the following things:
+
+1. Develop a comprehensive niche value check for all niches.
+2. Insert it ideally when values are loaded from memory (e.g. a "checked-load") or on function entry.
+3. Migrate the enum check to the same system mentioned in 2.
+
+### The "shiny future" we are working towards
+
+Similar to how [UBSan](https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html)
+exists in Clang, we would like to see an option in Rust to detect undefined
+behavior at runtime. As mentioned above, this is critical for cross-language
+interoperability and can help catch UB before it reaches production.
+
+The extension of these checks can be done step-by-step, keeping in mind the
+runtime overhead. Different checks can be enabled by separate flags (like in UBSan).
```

Comment:
```suggestion
The extension of these checks can be done step-by-step, keeping in mind the
overhead. Different checks can be enabled by separate flags (like in UBSan).
```
The compile-time overhead is more concerning to me. It's easy to double compile times with checks.

Reply:
Good point, thank you!